### ISSUE TITLE
Feature request: add optional tasks

### ISSUE THREAD

**Issue by nagromc** [anishathalye/dotbot#96]
I would like Dotbot to install the required softwares by my dotfiles. For instance, if I use `i3`, I would add to `install.conf.yaml`:

``` yaml
- shell:
    - sudo apt-get install -y i3

- link:
    - ~/.config/i3/config: i3/config
```

But imagine I need multiple softwares like `fish`, `tmux`, Vim, etc. and with multiple package managers (`npm install -g bower gulp yo`, `sudo pip install pre-commit pre-commit-hooks thefuck`, etc.), each time I run `./install`, the whole installation process will take several minutes to run, especially on low resource systems such as the Raspberry Pi v1.

I know Dotbot should be [idempotent](https://github.com/anishathalye/dotbot/blob/28959a3f31d8632975de4e18d0a0e47476b5a413/README.md#configuration), but wouldn't be great to run `./install --init` or something to execute specific part of the file (for instance, install the required softwares)? Then, the `install.conf.yaml` would become:

``` yaml
- optional:
    # Only executed if `--init` is specified
    - name: init
      tasks:
        - shell:
            - [sudo apt-get install -y i3, Installing required softwares]

# Executed in any case
- link:
    - ~/.config/i3/config: i3/config
```

**Comment by anishathalye** [comment]
Sounds interesting. I would be in favor of incorporating some form of this functionality.

This seems closely related to #44 / #81. It would be great to come up with a single design that unifies all of these ideas.

**Comment by nagromc** [comment]
Well, you're right. Hope we'll find a solution! It seems many of us would love such a feature.

**Comment by gwerbin** [comment]
Obviously @anishathalye you have the final word here, but I'd be concerned about feature creep. One of Dotbot's great strengths is that it's dead simple, and starting to add advanced configuration options might remove some of that simplicity.

You already have a plugin API -- I'd **much** rather see this kind of functionality implemented as a plugin than in the Dotbot core. Maybe there should be an interface for implementing new kinds of tasks; then there could be a "dotbot-optional-task" plugin that implements this feature.

**Comment by gwerbin** [comment]
However, letting tasks have some kind of `name` or `group` attribute seems like a useful feature to have in the core program. Then you can resolve https://github.com/anishathalye/dotbot/issues/81 easily by grouping the output together for tasks that share a `groups` attribute:

```yaml
- groups:
  - "init"
  shell:
    - [sudo apt-get install -y i3, Installing required softwares]

- link:
    - ~/.config/i3/config: i3/config
```

Alternatively you can require that all additional attributes have to be specified in an "attributes" element, which can then be passed on to plugins etc:

```yaml
- shell:
    - [sudo apt-get install -y i3, Installing required softwares]
  attributes:
  - optional
  - groups: ["init"]

- link:
    - ~/.config/i3/config: i3/config
```

**Comment by anishathalye** [comment]
I've merged in #214 to allow `--only` and `--except` to only run certain sections.

For groups of tasks, I think the right solution is splitting them into `foo.yaml`, `bar.yaml` and then tweaking the `install` script to take an argument that is the config name, e.g. `./install foo` which will call Dotbot on `foo.yaml`.
